Fix for #253

piponazo opened the pull request:
This PR should fix #253 and #246 

I still need to add the reproducer, but I would appreciate if you could take a look to the proposed solution and give some feedback.

The problem was that we were not checking if the marker was valid before calling `REPORT_MARKER`. This piece of code fixed the crash:

```
                if (marker != sos_) {
                    // Read the beginning of the next segment
                    marker = advanceToMarker();
                    if (marker < 0) {
                        throw Error(kerNoImageInInputData);
                    } else {
                        REPORT_MARKER;
                    }
                }
```
I also took the opportunity to apply clang-format in the affected function, so that the next time we have issues there we can find a code following our formatting rules.

piponazo committed:
clang format function

piponazo committed:
Add enforce before calling REPORT_MARKER with invalid marker

D4N commented on src/jpgimage.cpp:
```diff
@@ -795,7 +795,11 @@ namespace Exiv2 {
                     // Read the beginning of the next segment
                     marker = advanceToMarker();
-                    REPORT_MARKER;
+                    if (marker < 0) {
```
You know you could use `enforce()` for that ;-)

D4N reviewed (commented):
I think we can merge this once you add the reproducer.

piponazo commented:
I thought about it, but I did not try :). I will try it when I add the
reproducer

El vie., 30 mar. 2018 8:49, D4N <notifications@github.com> escribió:

> *@D4N* commented on this pull request.
>
> I think we can merge this once you add the reproducer.
> ------------------------------
>
> In src/jpgimage.cpp
> <https://github.com/Exiv2/exiv2/pull/256#discussion_r178241408>:
>
> > @@ -795,7 +795,11 @@ namespace Exiv2 {
>                  if (marker != sos_) {
>                      // Read the beginning of the next segment
>                      marker = advanceToMarker();
> -                    REPORT_MARKER;
> +                    if (marker < 0) {
>
> You know you could use enforce() for that ;-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Exiv2/exiv2/pull/256#pullrequestreview-108273255>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGQ9Tzw1vxvxKemGTtcg6qYa0jWreanks5tjdWPgaJpZM4TBWAT>
> .
>

D4N commented:
Don't feel pressured, I was just joking :)

piponazo committed:
Add reproducers for #253 and #246

piponazo commented:
@D4N I added the usage of `enforce` and added reproducers for #253 and #246 . However I am having issues that I do not understand with the latest one.

I get this error when trying to run the python test suite:

```
======================================================================
FAIL: test_run (bugfixes.github.test_issue_246.TestFirstPoC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/luis/programming/exiv2/tests/system_tests.py", line 521, in test_run
    _process_output_post(got_stdout.decode('utf-8')), stdout
  File "/home/luis/programming/exiv2/tests/system_tests.py", line 463, in compare_stdout
    self.assertMultiLineEqual(expected_stdout, got_stdout)
AssertionError: 'STRUCTURE OF JPEG FILE: /home/luis/progra[176 chars]00\n' != 'Usage: exiv2 [ options ] [ action ] file [44 chars]s.\n'
+ Usage: exiv2 [ options ] [ action ] file ...
+ 
+ Manipulate the Exif metadata of images.
- STRUCTURE OF JPEG FILE: /home/luis/programming/exiv2/test/data/1-string-format.jpg
-  address | marker       |  length | data
-        0 | 0xffd8 SOI  
-        2 | 0xffe1 APP1  |      60 | Exif..II*.....0.i...........0000
```
It's telling me that the stdout of the app is missing this first line: `Usage: exiv2 [ options ] [ action ] file ...`. However, when I run the application directly from the command line I get this:

```
✔ ~/programming/exiv2/build [fix253|✔] 
08:43 $ bin/exiv2 -pS ~/1-string-format.jpg 
STRUCTURE OF JPEG FILE: /home/luis/1-string-format.jpg
 address | marker       |  length | data
       0 | 0xffd8 SOI  
       2 | 0xffe1 APP1  |      60 | Exif..II*.....0.i...........0000
Exiv2 exception in print action for file /home/luis/1-string-format.jpg:
Input data does not contain a valid image
```
Do you know what am I missing?

D4N commented on tests/bugfixes/github/test_issue_246.py:
```diff
@@ -0,0 +1,25 @@
+    url = "https://github.com/Exiv2/exiv2/issues/246"
+
+    filename = "{data_path}/1-string-format.jpg"
+    commands = ["{exiv2} -pS" + filename]
```
you are missing a space after the "-pS", it should be:
``` Python
    commands = ["{exiv2} -pS " + filename]
```

D4N reviewed (approved):
Very good!

D4N commented:
Yes, the test suite has its rough points. I'll try to come up with something to reduce these kinds of mistakes (forgetting a space happens so often to me).
